systemtest/src/test/java/io/strimzi/systemtest/KafkaST.java

```diff
@@ -1671,6 +1675,145 @@ void testRegenerateCertExternalAddressChange() throws InterruptedException {
         assertNotEquals(secretsWithoutExt.getData().get("my-cluster-kafka-2.key"), secretsWithExt.getData().get("my-cluster-kafka-2.key"));
     }
 
+    @Test
+    void testLabelModificationDoesNotBreakCluster() throws Exception {
+        Map<String, String> labels = new HashMap<>();
+        String[] labelKeys = {"label-name-1", "label-name-2"};
+        String[] labelValues = {"name-of-the-label-1", "name-of-the-label-2"};
+        String[] labelNewValues = {"new-name-of-the-label-1", "new-name-of-the-label-2"};
+        String brokerServiceName = "my-cluster-kafka-brokers";
+        String configMapName = "my-cluster-kafka-config";
+
+        labels.put(labelKeys[0], labelValues[0]);
+        labels.put(labelKeys[1], labelValues[1]);
+
+        testMethodResources().kafkaEphemeral(CLUSTER_NAME, 3, 1)
+                .editSpec()
+                    .editKafka()
+                        .editListeners()
+                            .withNewKafkaListenerExternalNodePort()
+                            .endKafkaListenerExternalNodePort()
+                        .endListeners()
+                    .endKafka()
+                .endSpec()
+                .editMetadata()
+                    .withLabels(labels)
+                .endMetadata()
+                .done();
+
+        LOGGER.info("Waiting for kafka service labels changed {}", labels);
+        StUtils.waitForKafkaServiceLabelsChange(brokerServiceName, labels);
+
+        LOGGER.info("Getting labels from stateful set resource");
+        StatefulSet statefulSet = kubeClient().getStatefulSet(kafkaClusterName(CLUSTER_NAME));
+        labels = statefulSet.getSpec().getTemplate().getMetadata().getLabels();
+        LOGGER.info("Verifying default labels in the KafkaCR");
+
+        assertEquals(labelValues[0], labels.get(labelKeys[0]));
+        assertEquals(labelValues[1], labels.get(labelKeys[1]));
+
+        String newLabelKey = "label-name-3";
+        String newLabelValue = "name-of-the-label-3";
+
+        LOGGER.info("Edit kafka labels in Kafka CR");
+        replaceKafkaResource(CLUSTER_NAME, resource -> {
+            resource.getMetadata().getLabels().put(labelKeys[0], labelNewValues[0]);
+            resource.getMetadata().getLabels().put(labelKeys[1], labelNewValues[1]);
+            resource.getMetadata().getLabels().put(newLabelKey, newLabelValue);
+        });
+
+        labels.put(labelKeys[0], labelNewValues[0]);
+        labels.put(labelKeys[1], labelNewValues[1]);
+        labels.put(newLabelKey, newLabelValue);
+
+        LOGGER.info("Waiting for kafka service labels changed {}", labels);
+        StUtils.waitForKafkaServiceLabelsChange(brokerServiceName, labels);
+
+        LOGGER.info("Verifying kafka labels via services");
+        Service service = kubeClient().getService(brokerServiceName);
+
+        assertEquals(labelNewValues[0], service.getMetadata().getLabels().get(labelKeys[0]));
```

Comment:
All the resources you test - Service, Pods, StatefulSets, ConfigMap - can be treated as HasMetadata. I think you can write some support method for comparing the labels and just call it from here for each of the resources instead of having so many asserts. It will make the code easier to read.

Reply:
Thank you for your observation, edited.